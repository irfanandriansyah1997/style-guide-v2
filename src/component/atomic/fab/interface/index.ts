import { MouseEvent, ReactNode } from 'react';

import { IButtonThemeOption } from '@/atomic/button/interface';
import { IEventComponent } from '@/interface/general/event.interface';

export * from './fab-hooks.interface';

/**
 * Event Click Button
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.08
 */
export type IEventOnClickButton = IEventComponent<
  'on-click-button',
  MouseEvent<HTMLButtonElement>
>;

export type IFabAnimation =
  | 'fade-in'
  | 'fade-in-top'
  | 'fade-in-bottom'
  | 'fade-in-left'
  | 'fade-in-right';

export type IFabPosition = 'fixed' | 'absolute';

/**
 * Fab Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.13
 */
export interface IFabProps extends Partial<IFabStyling> {
  className?: string;
  icon: ReactNode;
  on(param: IEventOnClickButton): void;
  text?: string;
}

/**
 * Fab Styling Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.13
 */
export interface IFabStyling {
  animation: IFabAnimation;
  bottom: number | string;
  left: number | string;
  position: IFabPosition;
  right: number | string;
  rtl: boolean;
  shape: IFabType;
  show: boolean;
  theme: IButtonThemeOption;
  top: number | string;
}

export type IFabType = 'basic' | 'small' | 'extended';
