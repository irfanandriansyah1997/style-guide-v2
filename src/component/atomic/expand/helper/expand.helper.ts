import { objToString, verifiedIsNotEmpty } from '@99/helper';
import { Children, cloneElement, ReactElement, ReactNode } from 'react';

import {
  IExpandClassnameList as ClassnameList,
  IExpandClassnameList,
  IExpandContentProps,
  IExpandProps
} from '@/atomic/expand/interface';
import {
  ExpandContent,
  ExpandSelector
} from '@/atomic/expand/section/content.component';
import styles from '@/atomic/expand/style/style.module.scss';
import { checkClassnameAvailable } from '@/helper/component.helper';
import { PartialNull } from '@/interface/general';

interface IExpandChildren {
  content: ReactElement[];
  selector: ReactElement[];
}

/**
 * Generate Classname Container Element
 * @param {Partial<IExpandProps>} props - expand component props
 * @returns {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.06
 */
export const generateClassnameContainer = ({
  className,
  selectorPosition
}: Partial<IExpandProps>): string => {
  const response = {
    flex: true,
    'flex-column': selectorPosition === `top`,
    'flex-column-reverse': selectorPosition === `bottom`,
    [styles[`a-expand`]]: true,
    [`${className?.wrapper}`]: checkClassnameAvailable<ClassnameList>(
      className,
      `wrapper`
    )
  };

  return objToString(response);
};

/**
 * Transform Children To Expand Content
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.06
 */
export const generateContent = (
  children: ReactNode,
  classNameProps: PartialNull<IExpandClassnameList>
): ReactNode => {
  const response: ReactElement[] = [];
  const { content, selector } = Children.map(
    children as ReactElement,
    (child: ReactElement) => {
      const { type } = child as ReactElement;

      if ([ExpandContent, ExpandSelector].includes(type as any)) {
        return child;
      }

      return null;
    }
  )
    .filter(verifiedIsNotEmpty)
    .reduce(
      (prev: IExpandChildren, current): IExpandChildren => {
        const { type } = current as ReactElement;

        if ([ExpandContent].includes(type as any)) {
          return {
            ...prev,
            content: [...prev.content, current]
          };
        }

        return {
          ...prev,
          selector: [...prev.content, current]
        };
      },
      {
        content: [],
        selector: []
      }
    );

  if (selector.length > 0) response.push(selector[0]);
  if (content.length > 0) response.push(content[0]);

  return response.map((item, index) => {
    const { props } = item;
    const { className, ...res } = props as IExpandContentProps;

    return cloneElement(item, {
      ...res,
      className: objToString({
        [`${classNameProps?.toggle}`]:
          index === 0 &&
          checkClassnameAvailable<ClassnameList>(classNameProps, `toggle`),
        [`${classNameProps?.content}`]:
          index === 1 &&
          checkClassnameAvailable<ClassnameList>(classNameProps, `content`),
        [`${className}`]: verifiedIsNotEmpty(className)
      })
    });
  });
};
