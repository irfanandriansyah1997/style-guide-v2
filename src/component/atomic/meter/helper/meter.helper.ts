import { IMeterShape } from '@/atomic/meter/interface';

/**
 * Generate Rounded Type For Wrapper & Content Meter Component
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.19
 * @param {IMeterShape} shape - shape type meter component eg. rounded, shape or number for defined border radius size
 * @param {number}  height - height meter component
 * @returns {number}
 */
export const generateBorderRadiusMeter = (
  rounded: IMeterShape,
  height: number
): number => {
  switch (rounded) {
    case `rounded`:
      return height;

    case `square`:
      return 0;

    default:
      return rounded;
  }
};
