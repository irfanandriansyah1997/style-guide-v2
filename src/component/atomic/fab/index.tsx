import { objToString, verifiedIsNotEmpty } from '@99/helper';
import PropTypes from 'prop-types';
import { FC } from 'react';
import { Transition } from 'react-transition-group';

import { IButtonThemeOption } from '@/atomic/button/interface';
import styles from '@/atomic/fab/style/style.module.scss';
import Text from '@/atomic/text';

import { generateFABAnimation } from './helper/fab.helper';
import { useFab } from './hooks/fab.hooks';
import { IFabAnimation, IFabPosition, IFabProps, IFabType } from './interface';

/**
 * Floating Action Button Component
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.13
 */
const Fab: FC<IFabProps> = ({
  className: classNameProps,
  icon,
  on,
  text,
  ...styling
}) => {
  const { className, isExtended, style } = useFab(styling, classNameProps);

  return (
    <Transition in={styling.show} timeout={300}>
      {(state) => {
        const animationClassName = generateFABAnimation(
          styling.animation,
          state
        );

        return (
          <button
            style={style}
            type="button"
            onClick={(e) => {
              e.preventDefault();
              on({
                event: `on-click-button`,
                payload: e
              });
            }}
            className={objToString({
              [`${className}`]: true,
              [`${animationClassName}`]: verifiedIsNotEmpty(animationClassName)
            })}
          >
            <div
              className={objToString({
                flex: true,
                'flex-align-center': true,
                'flex-justify-center': true,
                [styles[`a-fab__icon`]]: true
              })}
            >
              {icon}
            </div>
            {isExtended && verifiedIsNotEmpty(text) && (
              <Text.Span
                className={objToString({
                  flex: true,
                  'flex-align-center': true,
                  'flex-justify-center': true,
                  [styles[`a-fab__text`]]: true
                })}
                font-weight={500}
              >
                {text}
              </Text.Span>
            )}
          </button>
        );
      }}
    </Transition>
  );
};

Fab.propTypes = {
  animation: PropTypes.oneOf<IFabAnimation>([
    `fade-in`,
    `fade-in-bottom`,
    `fade-in-left`,
    `fade-in-right`,
    `fade-in-top`
  ]),
  bottom: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  className: PropTypes.string,
  icon: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string
  ]).isRequired,
  left: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  on: PropTypes.func.isRequired,
  position: PropTypes.oneOf<IFabPosition>([`absolute`, `fixed`]),
  right: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  rtl: PropTypes.bool,
  shape: PropTypes.oneOf<IFabType>([`basic`, `extended`, `small`]),
  show: PropTypes.bool,
  text: PropTypes.string,
  theme: PropTypes.oneOf<IButtonThemeOption>([`dark`, `light`]),
  top: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Fab.defaultProps = {
  animation: `fade-in`,
  className: undefined,
  position: `absolute`,
  rtl: false,
  shape: `basic`,
  show: true,
  text: undefined,
  theme: `light`
};

export default Fab;
