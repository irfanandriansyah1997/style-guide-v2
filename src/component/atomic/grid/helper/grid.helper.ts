import { objToString, verifiedIsNotEmpty } from '@99/helper';
import { CSSProperties } from 'react';

import {
  IBasicSizeGrid,
  IGridRowStyling,
  IGridSize,
  IGridSpaceItem
} from '@/atomic/grid/interface';
import styles from '@/atomic/grid/style/style.module.scss';

/**
 * Generate Spacing Size
 * @param {IGridSpaceItem | number} param - spacing size props
 * @returns {IGridSpaceItem}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.16
 */
const generateSpacingSize = (
  param?: Partial<IGridSpaceItem> | number
): IGridSpaceItem => {
  const defaultSize = 12;

  if (!verifiedIsNotEmpty(param)) {
    return {
      horizontal: defaultSize,
      vertical: defaultSize
    };
  }

  if (typeof param === `number`) {
    return {
      horizontal: param / 2,
      vertical: param / 2
    };
  }

  return {
    horizontal: ((param as IGridSpaceItem).horizontal || defaultSize) / 2,
    vertical: ((param as IGridSpaceItem).vertical || defaultSize) / 2
  };
};

/**
 * Generate Class Name Wrapper
 * @param {string | undefined} className - custom classname
 * @returns {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.16
 */
export const generateClassNameItem = (
  size: IGridSize,
  { lg, md, sm, xl, xs }: Partial<IBasicSizeGrid>,
  className?: string
): string =>
  objToString({
    [`${className}`]: verifiedIsNotEmpty(className),
    [`${styles[`a-col`]}`]: size === `auto`,
    [`${styles[`a-col-${size}`]}`]: size !== `auto`,
    [`${styles[`a-col-xs-${xs}`]}`]: verifiedIsNotEmpty(xs) && xs !== `auto`,
    [`${styles[`a-col-xs`]}`]: verifiedIsNotEmpty(xs) && xs === `auto`,
    [`${styles[`a-col-sm`]}`]: verifiedIsNotEmpty(sm) && sm === `auto`,
    [`${styles[`a-col-sm-${sm}`]}`]: verifiedIsNotEmpty(sm) && sm !== `auto`,
    [`${styles[`a-col-md`]}`]: verifiedIsNotEmpty(md) && md === `auto`,
    [`${styles[`a-col-md-${md}`]}`]: verifiedIsNotEmpty(md) && md !== `auto`,
    [`${styles[`a-col-lg`]}`]: verifiedIsNotEmpty(lg) && lg === `auto`,
    [`${styles[`a-col-lg-${lg}`]}`]: verifiedIsNotEmpty(lg) && lg !== `auto`,
    [`${styles[`a-col-xl`]}`]: verifiedIsNotEmpty(xl) && xl === `auto`,
    [`${styles[`a-col-xl-${xl}`]}`]: verifiedIsNotEmpty(xl) && xl !== `auto`
  });

/**
 * Generate Class Name Wrapper
 * @param {string | undefined} className - custom classname
 * @returns {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.16
 */
export const generateClassNameWrapper = (className?: string): string =>
  objToString({
    flex: true,
    'flex-wrap': true,
    [`${className}`]: verifiedIsNotEmpty(className)
  });

/**
 * Generate Style Wrapper
 * @param {IGridRowStyling} props - props styling
 * @returns {CSSProperties}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.16
 */
export const generateStyleItem = (
  space: Partial<IGridSpaceItem> | number | undefined,
  order: number | undefined
): CSSProperties => {
  const { horizontal, vertical } = generateSpacingSize(space);

  return {
    order,
    paddingBottom: vertical,
    paddingLeft: horizontal,
    paddingRight: horizontal,
    paddingTop: vertical
  };
};

/**
 * Generate Style Wrapper
 * @param {IGridRowStyling} props - props styling
 * @returns {CSSProperties}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.16
 */
export const generateStyleWrapper = ({
  alignItems,
  justifyContent,
  spaceEachItem
}: IGridRowStyling): CSSProperties => {
  const { horizontal, vertical } = generateSpacingSize(spaceEachItem);

  return {
    alignItems,
    justifyContent,
    marginBottom: vertical,
    marginLeft: -1 * horizontal,
    marginRight: -1 * horizontal,
    marginTop: -1 * vertical
  };
};
