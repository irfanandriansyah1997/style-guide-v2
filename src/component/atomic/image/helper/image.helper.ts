import { verifiedIsNotEmpty, verifiedKeyIsExist } from '@99/helper';
import { ReactElement, ReactNode } from 'react';

import { IImagePresetProps } from '@/atomic/image/interface';
import { ImagePreset } from '@/atomic/image/section/image-responsive.component';
import { transformChildrenToArray } from '@/helper/component.helper';
import { IImageResponsive } from '@/interface/general';

/**
 * Transform Children To Image Responsive Object
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.06
 */
export const transformChildrenToImageResponsive = (
  children: ReactNode
): Partial<IImageResponsive> => {
  const props = transformChildrenToArray<IImagePresetProps>(children)(
    (child) => {
      if (
        verifiedIsNotEmpty(child) &&
        verifiedKeyIsExist((child as unknown) as Record<string, any>, `type`)
      ) {
        const { props, type } = child as ReactElement;

        if ([ImagePreset].includes(type as any)) {
          return props;
        }
      }

      return undefined;
    }
  ).sort((a, b) => a.sizes - b.sizes);

  return {
    src: props.length > 0 ? props[0].src : undefined,
    srcSet: (props || []).map(({ sizes, src }) => `${src} ${sizes}w`).join(`,`)
  };
};
