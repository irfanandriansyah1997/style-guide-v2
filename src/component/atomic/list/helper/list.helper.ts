import {
  objToString,
  verifiedIsNotEmpty,
  verifiedKeyIsExist
} from '@99/helper';
import { CSSProperties, Key, ReactElement, ReactNode } from 'react';

import {
  IListContent,
  IListContentType,
  IListItemProps,
  IListStyle
} from '@/atomic/list/interface';
import ListItem from '@/atomic/list/section/list-item.component';
import { transformChildrenToArray } from '@/helper/component.helper';

/**
 * Generate Classname Container
 * @param {Partial<IListStyle> | undefined} style - style object props
 * @returns {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.11
 */
export const generateClassnameContainer = (
  style: Partial<IListStyle> | undefined
): string => {
  let className: Record<string, boolean> = {
    flex: true,
    'flex-wrap': true,
    relative: true
  };
  if (style) {
    const { orientation } = style;

    switch (orientation) {
      case `horizontal`:
        className = {
          ...className,
          'flex-align-center': true,
          'flex-row': true
        };
        break;
      case `vertical`:
        className = {
          ...className,
          'flex-align-start': true,
          'flex-column': true
        };
        break;
    }
  }

  return objToString(className);
};

/**
 * Generate Style Item
 * @param {Partial<IListStyle> | undefined} style - style object props
 * @returns {CSSProperties}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.11
 */
export const generateStyleItem = (
  style: Partial<IListStyle> | undefined
): CSSProperties => {
  if (style) {
    const { dividerColor: borderColor, orientation, outline, space } = style;
    let padding: string | undefined;
    let borderStyle: string | undefined;

    if (space) {
      switch (orientation) {
        case `horizontal`:
          padding = `0px ${(space as number) / 2}px`;
          break;
        case `vertical`:
          padding = `${(space as number) / 2}px 0px`;
          break;
      }
    }

    if (outline) {
      switch (outline) {
        case `solid`:
          borderStyle = `solid`;
          break;
        case `dash`:
          borderStyle = `dashed`;
          break;
      }
    }

    return {
      borderColor,
      borderStyle,
      padding
    };
  }

  return {};
};

/**
 * Transform Children To List Item
 * @param {ReactNode} children - children Props
 * @returns {IBreadcrumbItemProps}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.06.11
 */
export const transformChildrenToListContent = (
  children: ReactNode
): IListContent[] =>
  transformChildrenToArray<{
    item: IListItemProps;
    key: Key | null;
  }>(children)((child) => {
    if (
      verifiedIsNotEmpty(child) &&
      verifiedKeyIsExist((child as unknown) as Record<string, any>, `type`)
    ) {
      const { key, props, type } = child as ReactElement;

      if ([ListItem].includes(type as any)) {
        return {
          item: props,
          key
        };
      }
    }

    return undefined;
  }).map(
    ({ item, key }, index): IListContent => ({
      key: key ? `${key}` : `${index}`,
      payload: item,
      type: IListContentType.item
    })
  );
