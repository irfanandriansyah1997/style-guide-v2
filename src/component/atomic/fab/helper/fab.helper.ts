import { objToString, verifiedIsNotEmpty } from '@99/helper';
import { CSSProperties } from 'react';
import { TransitionStatus } from 'react-transition-group';

import { IFabAnimation, IFabStyling } from '@/atomic/fab/interface';
import animationStyles from '@/atomic/fab/style/animation.module.scss';
import styles from '@/atomic/fab/style/style.module.scss';

/**
 * Generate FAB Animation while fade in & fade out
 * @param {IFabAnimation | undefined} animation - animation className
 * @param {TransitionStatus} state - transtion state
 * @return {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.14
 */
export const generateFABAnimation = (
  animation: IFabAnimation | undefined,
  state: TransitionStatus
): string => {
  let animationClassname = ``;
  if (verifiedIsNotEmpty(animation)) {
    switch (state) {
      case `entered`:
        animationClassname = `${animationStyles[`${animation}-entered`]}`;
        break;

      case `entering`:
        animationClassname = `${animationStyles[`${animation}-entering`]}`;
        break;

      case `exited`:
        animationClassname = `${animationStyles[`${animation}-exited`]}`;
        break;

      case `exiting`:
        animationClassname = `${animationStyles[`${animation}-exiting`]}`;
        break;

      default:
        throw new Error(`Unknown state`);
    }
  }

  return animationClassname;
};

/**
 * Generate FAB Classname
 * @param {IFabStyling} styling - fab styling props
 * @param {string | undefined} className - custom classname
 * @returns {string}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.14
 */
export const generateFABClassName = (
  { position, rtl, theme, type }: Partial<IFabStyling>,
  className: string | undefined
): string =>
  objToString({
    flex: true,
    'flex-align-center': true,
    'flex-justify-center': true,
    'flex-row': true,
    'flex-row-reverse': rtl,
    [`${position}`]: verifiedIsNotEmpty(position),
    [styles[`a-fab`]]: true,
    [styles[`a-fab--type-${type}`]]: true,
    [styles[`a-fab--theme-${theme}`]]: true,
    [`${className}`]: verifiedIsNotEmpty(className)
  });

/**
 * Generate FAB Style
 * @param {IFabStyling} styling - fab styling props
 * @returns {CSSProperties}
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.07.14
 */
export const generateFABStyle = ({
  bottom,
  left,
  position,
  right,
  top
}: Partial<IFabStyling>): CSSProperties => ({
  bottom,
  left,
  position,
  right,
  top
});
