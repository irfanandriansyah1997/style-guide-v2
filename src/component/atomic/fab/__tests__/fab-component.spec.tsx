import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import { TransitionStatus } from 'react-transition-group';

import Fab from '@/atomic/fab';
import { generateFABAnimation } from '@/atomic/fab/helper/fab.helper';

describe(`Testing fab component`, () => {
  describe(`Testing FAB Shape`, () => {
    it(`Testing Basic Shape With Light Theme`, () => {
      const onSpy = jest.fn();
      const { container } = render(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          shape="basic"
          show
          theme="light"
          position="absolute"
        />
      );

      expect(container.outerHTML).toMatchSnapshot();

      if (container.firstChild) {
        expect(container.firstChild).toHaveClass(`a-fab`);
        expect(container.firstChild).toHaveClass(`a-fab--shape-basic`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-small`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-extended`);
        expect(container.firstChild).toHaveClass(`a-fab--theme-light`);
        expect(container.firstChild).not.toHaveClass(`a-fab--theme-dark`);
        expect(container.firstChild).toHaveClass(`absolute`);
        expect(container.firstChild).not.toHaveClass(`fixed`);
        expect(container.firstChild).not.toHaveClass(`flex-row-reverse`);

        const { childNodes } = container.firstChild as HTMLElement;

        expect(childNodes.length).toBe(1);
        expect((childNodes[0] as HTMLElement).textContent).toBe(`★`);
      }
    });

    it(`Testing Small Shape With Dark Theme`, () => {
      const onSpy = jest.fn();
      const { container } = render(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          shape="small"
          show
          theme="dark"
          position="fixed"
        />
      );

      expect(container.outerHTML).toMatchSnapshot();

      if (container.firstChild) {
        expect(container.firstChild).toHaveClass(`a-fab`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-basic`);
        expect(container.firstChild).toHaveClass(`a-fab--shape-small`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-extended`);
        expect(container.firstChild).not.toHaveClass(`a-fab--theme-light`);
        expect(container.firstChild).toHaveClass(`a-fab--theme-dark`);
        expect(container.firstChild).not.toHaveClass(`absolute`);
        expect(container.firstChild).toHaveClass(`fixed`);
        expect(container.firstChild).not.toHaveClass(`flex-row-reverse`);

        const { childNodes } = container.firstChild as HTMLElement;

        expect(childNodes.length).toBe(1);
        expect((childNodes[0] as HTMLElement).textContent).toBe(`★`);
      }
    });

    it(`Testing Extended Shape With Dark Theme`, () => {
      const onSpy = jest.fn();
      const { container } = render(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          text="Sample Text"
          shape="extended"
          show
          theme="dark"
          rtl
          position="fixed"
        />
      );

      expect(container.outerHTML).toMatchSnapshot();

      if (container.firstChild) {
        expect(container.firstChild).toHaveClass(`a-fab`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-basic`);
        expect(container.firstChild).not.toHaveClass(`a-fab--shape-small`);
        expect(container.firstChild).toHaveClass(`a-fab--shape-extended`);
        expect(container.firstChild).not.toHaveClass(`a-fab--theme-light`);
        expect(container.firstChild).toHaveClass(`a-fab--theme-dark`);
        expect(container.firstChild).not.toHaveClass(`absolute`);
        expect(container.firstChild).toHaveClass(`fixed`);
        expect(container.firstChild).toHaveClass(`flex-row-reverse`);

        const { childNodes } = container.firstChild as HTMLElement;

        expect(childNodes.length).toBe(2);
        expect((childNodes[0] as HTMLElement).textContent).toBe(`★`);
        expect((childNodes[1] as HTMLElement).textContent).toBe(`Sample Text`);
      }
    });
  });

  describe(`Testing FAB Shape`, () => {
    it(`Customize Position`, () => {
      const onSpy = jest.fn();
      const { container } = render(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          text="Sample Text"
          shape="extended"
          show
          theme="dark"
          rtl
          position="fixed"
          left={10}
          right={10}
          bottom={10}
          top={10}
          animation="fade-in"
        />
      );

      expect(container.outerHTML).toMatchSnapshot();

      if (container.firstChild) {
        expect(container.firstChild).toHaveStyle({ bottom: `10px` });
      }
    });
  });

  describe(`Testing FAB Animation`, () => {
    it(`Set Show`, () => {
      const onSpy = jest.fn();
      const component = mount(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          text="Sample Text"
          shape="extended"
          theme="dark"
          rtl
          show={false}
          position="fixed"
          left={10}
          right={10}
          bottom={10}
          top={10}
          animation="fade-in"
        />
      );

      component.setProps({
        show: true
      });
      component.update();

      expect(
        component.find(`button`).hasClass(`fade-in-entering`)
      ).toBeTruthy();
      expect(component.find(`button`).hasClass(`fade-in-exiting`)).toBeFalsy();
    });

    it(`Set Hidden`, () => {
      const onSpy = jest.fn();
      const component = mount(
        <Fab
          icon={<div>★</div>}
          on={onSpy}
          text="Sample Text"
          shape="extended"
          theme="dark"
          rtl
          show
          position="fixed"
          left={10}
          right={10}
          bottom={10}
          top={10}
          animation="fade-in"
        />
      );

      component.setProps({
        show: false
      });
      component.update();

      expect(component.find(`button`).hasClass(`fade-in-entering`)).toBeFalsy();
      expect(component.find(`button`).hasClass(`fade-in-exiting`)).toBeTruthy();
    });

    it(`Pass wrong parameter when generate className`, () => {
      try {
        generateFABAnimation(
          `fade-in`,
          (`unknown` as unknown) as TransitionStatus
        );
      } catch (e) {
        if (e instanceof Error) {
          expect(e.message).toBe(`Unknown state`);
        }
      }
    });
  });
});
