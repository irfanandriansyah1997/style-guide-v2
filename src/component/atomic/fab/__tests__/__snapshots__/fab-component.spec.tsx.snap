// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing fab component Testing FAB Shape Customize Position 1`] = `"<div><button style=\\"bottom: 10px; left: 10px; right: 10px; top: 10px;\\" type=\\"button\\" class=\\"flex flex-align-center flex-justify-center flex-row flex-row-reverse fixed a-fab a-fab--shape-extended a-fab--theme-dark fade-in-entered\\"><div class=\\"flex flex-align-center flex-justify-center a-fab__icon\\"><div>★</div></div><span font-weight=\\"500\\" class=\\"a-text flex flex-align-center flex-justify-center a-fab__text\\">Sample Text</span></button></div>"`;

exports[`Testing fab component Testing FAB Shape Testing Basic Shape With Light Theme 1`] = `"<div><button type=\\"button\\" class=\\"flex flex-align-center flex-justify-center flex-row absolute a-fab a-fab--shape-basic a-fab--theme-light fade-in-entered\\"><div class=\\"flex flex-align-center flex-justify-center a-fab__icon\\"><div>★</div></div></button></div>"`;

exports[`Testing fab component Testing FAB Shape Testing Extended Shape With Dark Theme 1`] = `"<div><button type=\\"button\\" class=\\"flex flex-align-center flex-justify-center flex-row flex-row-reverse fixed a-fab a-fab--shape-extended a-fab--theme-dark fade-in-entered\\"><div class=\\"flex flex-align-center flex-justify-center a-fab__icon\\"><div>★</div></div><span font-weight=\\"500\\" class=\\"a-text flex flex-align-center flex-justify-center a-fab__text\\">Sample Text</span></button></div>"`;

exports[`Testing fab component Testing FAB Shape Testing Small Shape With Dark Theme 1`] = `"<div><button type=\\"button\\" class=\\"flex flex-align-center flex-justify-center flex-row fixed a-fab a-fab--shape-small a-fab--theme-dark fade-in-entered\\"><div class=\\"flex flex-align-center flex-justify-center a-fab__icon\\"><div>★</div></div></button></div>"`;
